### PR TITLE
Add Sonoma Data Scraper

### PR DESCRIPTION
**Pull request by ldtcooper**

### CODE REVIEW

**Comment by ldtcooper** [comment]
The code is mostly done now. At this point, I'm just doing some clean-up and clarification, hence the draft status. 
I know my implementation is somewhat repetitive, and I'm open to suggestions to fix that. I played around with the idea of having a generic function to get the rows and cells from a tag, loop through them, and take a function to do any extra cleaning/transformation on them, but that made things more complicated and hard-to-read IMO.

File: data_scrapers/sonoma_county.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        transmissions[type] = int(number)
+    return transmissions
+
+def transform_tests(tests_tag: element.Tag) -> Dict[str, int]:
+    tests = {}
+    rows = get_rows(tests_tag)
+    for row in rows:
+        result, number, _pct = get_cells(row)
+        lower_res = result.lower()
+        tests[lower_res] = int(number.replace(',', ''))
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:102]: It feels like a helper for parsing integers might be useful:

```py
def parse_int(text: str) -> int:
    text = text.strip()
    if text == '-':
        return 0
    else:
        return int(text.replace(',', ''))
```

File: data_scrapers/sonoma_county.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+    """
+    Transform function for tables which don't require any special processing.
+    Takes in a BeautifulSoup tag for a table and returns a dictionary
+    in which the keys are strings and the values integers
+    """
+    categories = {}
+    rows = get_rows(tag)
+    for row in rows:
+        cat, cases, _pct = get_cells(row)
+        categories[cat] = int(cases)
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:115]: Hopefully it’s never an issue, but I’ll bet you’re going to need similar parsing logic as above with commas if this ever grows big enough.

File: data_scrapers/sonoma_county.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    return categories
+
+def get_unknown_race(race_eth_tag: element.Tag) -> int:
+    """
+    Gets the notes under the 'Cases by race and ethnicity' table to find the
+    number of cases where the person's race is unknown
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:121]: Oh, Sonoma, you could have just put it in the table. 😐

File: data_scrapers/sonoma_county.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    """
+    race_cases = {}
+    race_transform = {'Asian/Pacific Islander, non-Hispanic': 'Asian', 'Hispanic/Latino': 'Latinx_or_Hispanic', 'Other*, non-Hispanic': 'Other', 'White, non-Hispanic': 'White'}
+    rows = get_rows(race_eth_tag)
+    for row in rows:
+        group_name, cases, _pct = get_cells(row)
+        if group_name not in race_transform:
+            raise FutureWarning('The racial group {0} is new in the data -- please adjust the scraper accordingly')
+        internal_name = race_transform[group_name]
+        race_cases[internal_name] = int(cases)
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:145]: Same here with int parsing — this will probably break with values > 999.

File: data_scrapers/sonoma_county.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    return(int(matches.groups()[0]))
+
+def transform_race_eth(race_eth_tag: element.Tag) -> Dict[str, int]:
+    """
+    Takes in the BeautifulSoup tag for the cases by race/ethnicity table and
+    transforms it into an object of form:
+    'race_eth': {'Asian': -1, 'Latinx_or_Hispanic': -1, 'Other': -1, 'White':-1, 'Unknown': -1}
+    NB: These are the only races reported seperatley by Sonoma county at this time
+    """
+    race_cases = {}
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:137]: In Alameda, I believe @elaguerta loaded the defaults from the data model JSON so that any race/ethnicities we don’t have values for come through as `-1` or `0` instead of being unlisted. We should *probably* do the same here.

File: data_scrapers/sonoma_county.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    """
+    Takes in a BeautifulSoup tag of the cases by hospitalization table and
+    returns a dictionary with the numbers of hospitalized and non-hospitalized
+    cases
+    """
+    hospitalizations = {}
+    rows = get_rows(hospital_tag)
+    for row in rows:
+        hospitalized, number, _pct = get_cells(row)
+        if hospitalized == 'Yes':
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:159]: Maybe worth a case-insensitive comparison, just in case this changes in the future.

File: data_scrapers/sonoma_county.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    Takes in a BeautifulSoup tag representing the percent of cases hospitalized
+    by gender and returns a dictionary of those percentages in float form
+    e.g. 9% is 0.09
+    """
+    hospitalized = {}
+    rows = get_rows(hospital_tag)
+    for row in rows:
+        gender, no, yes = get_cells(row)
+        yes_int = int(yes.replace('%', ''))
+        hospitalized[gender] = (yes_int / 100)
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:176]: Would it be useful to lower-case gender here?

File: data_scrapers/sonoma_county.py [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    return hospitalized
+
+def get_county() -> Dict:
+    """Main method for populating county data .json"""
+    url = 'https://socoemergency.org/emergency/novel-coronavirus/coronavirus-cases/'
+    page = requests.get(url)
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:182]: Might be useful to call `page.raise_for_status()` just in case you got an error page.

File: data_scrapers/sonoma_county.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    sonoma_soup = BeautifulSoup(page.content, 'html.parser')
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:183]: Might be good to use `html5lib` instead of `html.parser`.

File: data_scrapers/sonoma_county.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    tables = sonoma_soup.findAll('table')[4:] # we don't need the first three tables
+
+    try:
+        # we have a lot more data here than we are using
+        hist_cases, cases_by_source, cases_by_race, total_tests, cases_by_region, region_guide, hospitalized, underlying_cond, symptoms, cases_by_gender, underlying_cond_by_gender, hospitalized_by_gender, symptoms_female, symptoms_male, symptoms_desc, cases_by_age, symptoms_by_age, underlying_cond_by_age = tables
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:188]: This is super convenient, but I worry about the order changing here. Maybe safer to call `sonoma_soup.find(re.compile(r'^h\d$'), string='whatever')` for each?

File: data_scrapers/sonoma_county.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+    model = {
+        'name': 'Sonoma County',
+        'update_time': generate_update_time(sonoma_soup),
+        'source': url,
+        'meta_from_source': get_source_meta(sonoma_soup),
+        'meta_from_baypd': 'Racial "Other" category includes "Black/African American, American Indian/Alaska Native, and Other"',
+        'series': transform_cases(hist_cases),
+        'case_totals': {
+            'transmission_cat': transform_transmission(cases_by_source),
+            'age_group': generic_transform(cases_by_age),
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:201]: This gets us the wrong format:

```js
        "age_group": {
            "0-17": 58,
            "18-49": 190,
            "50-64": 78,
            "65 and Above": 47,
            "Under Investigation": 0
        },
```

But we should have a list, and the group names should be slightly different: https://github.com/sfbrigade/data-covid19-sfbayarea/tree/master/data_models#cases-2

```js
[
	{"group": "0_to_17", "raw_count": -1 },
	{"group": "18_to_49", "raw_count": -1 },
	{"group": "50_to_64", "raw_count": -1 },
	{"group": "65_and_older", "raw_count": -1 },
	{"group": "Unknown", "raw_count": -1 }
]
```

File: data_scrapers/sonoma_county.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+            'race_eth': transform_race_eth(cases_by_race),
+            'gender': generic_transform(cases_by_gender)
```
Comment (Mr0grog) [inline data_scrapers/sonoma_county.py:203]: This is outputting different keys than we have for hospitalizations below. We should unify them.

Here it’s plural:

```js
        "gender": {
            "Males": 178,
            "Females": 195
        }
```

In hospitalizations it’s singular:

```js
    "hospitalizations": {
        "hospitalized_cases": {
            "hospitalized": 35,
            "not_hospitalized": 338
        },
        "gender": {
            "Male": 0.08,
            "Female": 0.1
        }
    }
```

File: data_scrapers/sonoma_county.py (hunk 11 above)
Comment (elaguerta) [reply data_scrapers/sonoma_county.py:201]: Our data_model.json had inconsistent use of "age_group" values (only 1 instance was a list). I changed all the "age_group" tables to be lists in [#52, in progress](https://github.com/sfbrigade/data-covid19-sfbayarea/pull/52). 
As far as using the group names, I realized I need to go do that for Alameda County and SF county. And we will probably have to manually order the list by group-name, since they're often coming through as unordered .json objects (in curly braces).

File: data_scrapers/sonoma_county.py (hunk 6 above)
Comment (elaguerta) [reply data_scrapers/sonoma_county.py:159]: I thought we would use the [chhs](https://data.chhs.ca.gov/dataset/california-covid-19-hospital-data-and-case-statistics/resource/6cd8d424-dfaa-4bdd-9410-a3d656e1176e?inner_span=True) as a common source for all hospitalization records. It looks like this is where most counties are getting their information, and we can be sure to have some sort of standard across counties by going straight to the chhs API. This referenced in issue #29. This is why the data_model doesn't currently include a hospitalization sub-table, I thought we could add it separately from the individual county scrapers. I decided to exclude hospitalization data available through SF county, to avoid confusion when we do get to #29.

File: README.md [hunk 13]
```diff
@@ -9,4 +9,158 @@ To install this project, you can simply run `sh install.sh` in your terminal. Th
 To run the scraper, you can use the run script by typing `sh run_scraper.sh` into your terminal. This will enable the virtual environment and run `scraper.py`. Once again, the virtual environment will not stay active after the script finishes running. If you want to run the scraper without the run script, enable the virtual environment, then run `python3 scraper.py`.
 
 ## Running the API
-The best way to run the API right now is to run the command `FLASK_APP="app.py" FLASK_ENV=development flask run;`. Note that this is not the best way to run the scraper at this time.
\ No newline at end of file
+The best way to run the API right now is to run the command `FLASK_APP="app.py" FLASK_ENV=development flask run;`. Note that this is not the best way to run the scraper at this time.
+
+## Data Model
```
Comment (elaguerta) [inline README.md:14]: Do you want to merge the .data_scrapers/README  into the top level README? See PR #48

File: data_scrapers/sonoma_county.py [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+        new_deaths = dead - cumul_deaths
+        deaths.append({ 'date': date, 'deaths': new_deaths, 'cumul_deaths': dead })
+
+        new_recovered = recoveries - cumul_recovered
+        recovered.append({ 'date': date, 'recovered': new_recovered, 'cumul_recovered': recoveries })
+
+        new_active = active_cases - cumul_active
+        active.append({ 'date': date, 'active': new_active, 'cumul_active': active_cases })
+
+    return { 'cases': cases, 'deaths': deaths, 'recovered': recovered, 'active': active }
```
[the remark this answers is not in the document]
Comment (elaguerta) [reply data_scrapers/sonoma_county.py:76]: Do we want to include "active" and "recovered" time-series even if they're not in the data model? I believe Sonoma is the only county with these.  I'm personally suspicious of the label "recovered" when things like strokes are affecting people long after they test as negative. We're also excluding all the detailed breakdowns Sonoma has over symptoms and underlying conditions, which makes me ask the question: if data isn't in our data model, are there criteria we're applying to decide whether to include or exclude it?

[the remark this answers is not in the document]
Comment (ldtcooper) [reply data_scrapers/sonoma_county.py:76]: I put active and recovered in because they were there and I'd rather err on the side of putting too much data in. If we don't want to use them, then I'll take them out of the output.

File: data_scrapers/sonoma_county.py [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+    cumul_deaths = 0
+    recovered = []
+    cumul_recovered = 0
+    active = []
+    cumul_active = 0
+    rows = get_rows(cases_tag)
+    for row in rows:
+        row_cells = row.findAll(['th', 'td'])
+        # print(type(row_cells))
+        date = row_cells[0].text.replace('/', '-')
```
[the remark this answers is not in the document]
Comment (ldtcooper) [reply data_scrapers/sonoma_county.py]: All of the dates are now using `dateutils`

File: data_scrapers/sonoma_county.py [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+    {'community': -1, 'from_contact': -1, 'travel': -1, 'unknown': -1}
+    """
+    transmissions = {}
+    rows = get_rows(transmission_tag)
+    # turns the transmission categories on the page into the ones we're using
+    transmission_type_conversion = {'Community': 'community', 'Close Contact': 'from_contact', 'Travel': 'travel', 'Under Investigation': 'unknown'}
+    for row in rows:
+        type, number, _pct = get_cells(row)
+        if type not in transmission_type_conversion:
+            raise FutureWarning('The transmission type {0} was not found in transmission_type_conversion'.format(type))
```
[the remark this answers is not in the document]
Comment (ldtcooper) [reply data_scrapers/sonoma_county.py]: I created a custom `FormatError` exception we can use throughout the project

File: data_scrapers/sonoma_county.py (hunk 3 above)
Comment (ldtcooper) [reply data_scrapers/sonoma_county.py:144]: 🤷

**Comment by ldtcooper** [comment]
Alright @Mr0grog I think I'm ready for a second round of review

File: data_scrapers/sonoma.py [hunk 17]
```diff
@@ -0,0 +1,299 @@
...
+    for row in rows:
+        gender, cases, _pct = get_cells(row)
+        if gender not in gender_string_conversions:
+            raise FormatError('An unrecognized gender has been added to the gender table')
+        categories[gender_string_conversions[gender]] = parse_int(cases)
+    return categories
+
+# not sure why I need object in here: I've checked the output value types and
+# they are strings and ints, but mypy is convinced there is an object in there
+def age_transform(tag: element.Tag) -> List[Dict[str, Union[str, int, object]]]:
```
Comment (Mr0grog) [inline data_scrapers/sonoma.py:159]: Minor nit: should use the same naming convention as the other transform functions, which are all `def transform_xyz`.

> not sure why I need object in here: I've checked the output value types and they are strings and ints, but mypy is convinced there is an object in there

Oh, this is super interesting! I think it has to do with [how Mypy uses contextual info](https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html#context-in-type-inference). For example, this is OK:

```py
def return_union() -> Dict[str, Union[str, int]]:
    a = 'hi'
    b = 5
    return {'a': a, 'b': b}
```

But this causes the problem you’re seeing:

```py
def return_union() -> Dict[str, Union[str, int]]:
    a = 'hi'
    b = 5
    result = {'a': a, 'b': b}
    return result
```

Based on the above docs, I’m all but certain what’s happening is that when Mypy sees:

```py
{'a': a, 'b': b}
# Or in our case, {'group': age_string_transform[group], 'raw_cases': raw_cases}
```

…it has to infer the correct value type. There are a *lot* of valid ways to do that, and what it goes for is the most common super-type of all the values, which is `object` (both `int` and `str` are subtypes of `object`, as is basically everything — it’s pretty similar to `Any`). So the type Mypy infers is `Dict[str, object]`.

BUT when it sees:

```py
return {'a': a, 'b': b}
```

It’s contextually aware enough to know that, because this is being returned, it should go look at the declared return value and see if that would *also* be a compatible alternative. It is, so Mypy just infers that the type of the expression is the one you declared (`Dict[str, Union[str, int]]`).

In the former case, where there wasn’t an immediate return in front of the dict expression, Mypy has no way to tell that it should compare it with the declared return type to see if that’s a better alternative. Then by the time it gets to a return statement, all it knows is that all the dicts you put in the `categories` list are of type `Dict[str, object]`.

**Anyway, long story short, there are two options that can fix this:**

1. When you declare `categories`, declare its type, and append the dict expression to it instead of a variable with the dict (it’s looking at the context again):

    ```py
    categories: List[Dict[str, Union[str, int]]] = []
    ...
        categories.append({'group': age_string_transform[group],
                           'raw_cases': raw_cases})
    ```

2. Declare the dict type before appending it down on line 181:

    ```py
    # Give Mypy a better type to guess than Dict[str, object]
    element: Dict[str, Union[str, int]] = {
        'group': age_string_transform[group],
        'raw_cases': raw_cases
    }
    categories.append(element)
    ```

File: data_scrapers/sonoma.py [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+    returns a dictionary with the numbers of hospitalized and non-hospitalized
+    cases
+    """
+    hospitalizations = {}
+    rows = get_rows(hospital_tag)
+    for row in rows:
+        hospitalized, number, _pct = get_cells(row)
+        if hospitalized.lower() == 'yes':
+            hospitalizations['hospitalized'] = parse_int(number)
+        else:
```
Comment (Mr0grog) [inline data_scrapers/sonoma.py:238]: Might be worth checking the value here anyway. (And also raising an error is `hospitalized.lower()` is something totally unexpected.) What if the row titles change to `Hospitalized` and `Not Hospitalized`? (Or something else.) Then we’d wind up with:

```js
{
  // Note "hospitalized" is not even present.
  "not_hospitalized": <value of last row in table, regardless of what it represents>
}
```

File: data_scrapers/sonoma.py [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+
+def get_county() -> Dict:
+    """
+    Main method for populating county data .json
+    """
+    url = 'https://socoemergency.org/emergency/novel-coronavirus/coronavirus-cases/'
+    page = requests.get(url)
+    page.raise_for_status()
+    sonoma_soup = BeautifulSoup(page.content, 'html5lib')
+    tables = sonoma_soup.find_all('table')[4:] # we don't need the first three tables
```
Comment (Mr0grog) [inline data_scrapers/sonoma.py:271]: Looks like this line is vestigial.

File: data_scrapers/sonoma.py [hunk 20]
```diff
@@ -0,0 +1,299 @@
...
+    cumul_cases = 0
+    deaths = []
+    cumul_deaths = 0
+    # recovered = []
+    # cumul_recovered = 0
+    # active = []
+    # cumul_active = 0
+    rows = get_rows(cases_tag)
+    for row in rows:
+        row_cells = row.find_all(['th', 'td'])
```
[the remark this answers is not in the document]
Comment (Mr0grog) [reply data_scrapers/sonoma.py:79]: To be clear, it’s cool if the answer is “tried it, but too much of a pain/not worth the tradeoff.” I just want to make sure it didn’t get missed entirely since I’m concerned about changing columns.

File: data_scrapers/sonoma.py [hunk 21]
```diff
@@ -0,0 +1,299 @@
...
+        raise ValueError(f'Date is not in ISO 8601'
+                         f'format: "{update_time_text}"')
+    return date.isoformat()
+
+def get_source_meta(soup: BeautifulSoup) -> str:
+    """
+    Finds the 'Definitions' header on the page and gets all of the text in it
+    """
+    definitions_header = soup.find('h3', string='Definitions')
+    definitions_text = definitions_header.find_parent().text
```
[the remark this answers is not in the document]
Comment (ldtcooper) [reply data_scrapers/sonoma.py:56]: Yes, good catch!

File: data_scrapers/sonoma.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (ldtcooper) [reply data_scrapers/sonoma.py:79]: Oh, sorry, I missed that suggestion completely. I'll give it a try

[the remark this answers is not in the document]
Comment (ldtcooper) [reply data_scrapers/sonoma.py]: I started trying this, and I think it's going to be more work than is worth it right now. The shape of the tables has been consistent since I started working on this, although the order of the tables on the page has changed. The function to parse tables is in there, and we can switch over to using it if the table format starts changing

File: data_scrapers/sonoma.py [hunk 22]
```diff
@@ -38,6 +38,22 @@ def get_cells(row: element.ResultSet) -> List[str]:
     """
     return [el.text for el in row.find_all(['th', 'td'])]
 
+def row_list_to_dict(row_list: List[str], headers: List[str]) -> Dict[str, str]:
+    output = {}
+    for i in range(len(row_list)):
+        val = row_list[i]
+        header = headers[i]
+        output[header] = val
+    return output
```
Comment (Mr0grog) [inline data_scrapers/sonoma.py:47]: This whole function can actually just be done with:

```py
dict(zip(headers, row_list))
```

So you could simplify this function to:

```py
def row_list_to_dict(row_list, headers):
    return dict(zip(headers, row_list))
```

Or get rid of it entirely and replace line 55 with:

```py
return [dict(zip(header_cells, row)) for row in body_cells]
```

File: data_scrapers/sonoma.py [hunk 23]
```diff
@@ -38,6 +38,22 @@ def get_cells(row: element.ResultSet) -> List[str]:
...
+
+def parse_table(tag: element.Tag) -> List[Dict[str, str]]:
+    rows = tag.find_all('tr')
+    header = rows[0]
+    body = rows[1:]
+    header_cells = get_cells(header)
+    body_cells = [get_cells(row) for row in body]
```
Comment (Mr0grog) [inline data_scrapers/sonoma.py:54]: It’s a minor performance improvement, but you could make this a generator comprehension:

```suggestion
    body_cells = (get_cells(row) for row in body)
```

Less memory (no extra list is created for `body_cells`) and fewer iterations, so less CPU, too.

File: data_scrapers/sonoma.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (Mr0grog) [reply data_scrapers/sonoma.py]: OK.

> we can switch over to using it if the table format starts changing

But just to be clear, part of what I was getting at is that it might be hard to notice that the format has started changing. With code like this:

```py
def transform_cases(...):
    ...
    active_cases, new_infected, dead, recoveries = [parse_int(el.text) for el in row_cells[1:]]
    ....
```

If those columns continue to be numeric, nothing about this program will tell you that you’re getting the wrong values out when things were rearranged, new columns were inserted, etc.

[the remark this answers is not in the document]
Comment (Mr0grog) [reply data_scrapers/sonoma.py]: (This would also help alert you to the fact that the order of tables has changed, since the re-ordered tables would presumably have different headers than you expected, causing errors when you try to get cells by their header.)

File: data_scrapers/sonoma.py [hunk 24]
```diff
@@ -0,0 +1,324 @@
...
+def transform_gender_hospitalizations(hospital_tag: element.Tag) -> Dict[str, float]:
+    """
+    Takes in a BeautifulSoup tag representing the percent of cases hospitalized
+    by gender and returns a dictionary of those percentages in float form
+    e.g. 9% is 0.09
+    """
+    hospitalized = {}
+    rows = get_rows(hospital_tag)
+    for row in rows:
+        gender, no, yes = get_cells(row)
```
[the remark this answers is not in the document]
Comment (ldtcooper) [reply data_scrapers/sonoma.py]: I ended up taking out the hospitalization code. I want to start looking into refactoring it, but I'd rather get a working version out before that.

File: README.md (hunk 13 above)
Comment (Mr0grog) [reply README.md:14]: @ldtcooper any thoughts on this point from @elaguerta? There are a lot of great improvements in #48 and it seems like we should have merged it quite a while ago.

This is kind of in conflict with that. I think we should merge #48. I’m not sure whether you’re thinking we should also move all that content to the top level README (seems pretty big and maybe a bit unwieldy, but that is not a strongly held opinion).

**Comment by elaguerta** [comment]
Heads up, I also just merged #48 so that may conflict with your `README.md` s

**Comment by ldtcooper** [comment]
It looks like it does. I'll try to fix those conflicts and linter errors tonight

**Comment by ldtcooper** [comment]
Looks like all the linter and merge problems have been fixed @Mr0grog @elaguerta

File: README.md (hunk 13 above)
Comment (Mr0grog) [reply README.md:14]: So this now repeats a lot of information that is already in [`data_models/README.md`](https://github.com/sfbrigade/data-covid19-sfbayarea/tree/sonoma/data_models#age-group). Would it be better to just link to it? That way we don’t get in a situation where someone updates one place but not the other.

File: covid19_sfbayarea/data/sonoma.py [hunk 25]
```diff
@@ -0,0 +1,273 @@
+#!/usr/bin/env python3
+import requests
+import json
+import re
+import dateutil.parser
+from typing import List, Dict, Union
+from bs4 import BeautifulSoup, element # type: ignore
+from format_error import FormatError # type: ignore
```
Comment (Mr0grog) [inline covid19_sfbayarea/data/sonoma.py:8]: This is broken for me. If I run `./run_scraper_data.sh solano` or `python scraper_data.py solano`, I get:

```
Traceback (most recent call last):
  File "scraper_data.py", line 4, in <module>
    from covid19_sfbayarea import data as data_scrapers
  File "/Users/rbrackett/Dev/sfbrigade/data-covid19-sfbayarea/covid19_sfbayarea/data/__init__.py", line 4, in <module>
    from . import sonoma
  File "/Users/rbrackett/Dev/sfbrigade/data-covid19-sfbayarea/covid19_sfbayarea/data/sonoma.py", line 8, in <module>
    from format_error import FormatError
ModuleNotFoundError: No module named 'format_error'
```

Does it work for you? Seems like it should be:

```suggestion
from .format_error import FormatError
```

(It also shouldn’t need need the `# type: ignore` comment.)

---

Also, while you’re here, maybe it’s worthwhile to make this a shared module up in `covid19_sfbayarea/errors.py` instead of `covid19_sfbayarea/data/format_error.py`? Some of the news scrapers also use [an identical class](https://github.com/sfbrigade/data-covid19-sfbayarea/blob/47ced984c3379fa4b5908e3de1e9778714d5e7c4/covid19_sfbayarea/news/errors.py#L1-L7), and we should really only have one implementation. Then this line would be:

```suggestion
from ..errors import FormatError
```

File: README.md (hunk 13 above)
Comment (ldtcooper) [reply README.md:14]: I thought we were merging `data_models/README.md` into the main readme

Comment (Mr0grog) [reply README.md:14]: OK! I was unclear on exactly which direction we were going there. If we’re merging into the top-level README, then this is missing a bunch of stuff from `data_models/README.md`, and this PR should probably also remove that file.

**Comment by elaguerta** [comment]
They are using ArcGIS dashboards now. You can look at Alameda as a model. 
I found some endpoints.

- **Cases and deaths reported by Date:** https://services1.arcgis.com/P5Mv5GY5S66M8Z1Q/arcgis/rest/services/NCOV_Cases_Sonoma_County/FeatureServer
- **Hodge Podge of Tables Appended Together, including Age Group, Gender, Test Results, and Race Ethnicity:** https://services1.arcgis.com/P5Mv5GY5S66M8Z1Q/ArcGIS/rest/services/NCOV_Cases_Sonoma_County_Statistics/FeatureServer

You might find this post helpful: https://stackoverflow.com/questions/50161492/how-do-i-scrape-data-from-an-arcgis-online-map

Also, we may want to address #85 now that there are 2 counties using ArcGIS.

File: covid19_sfbayarea/data/sonoma.py [hunk 26]
```diff
@@ -0,0 +1,261 @@
...
+    race_cases = {
+        'African_Amer': 0,
+        'Asian': 0,
+        'Latinx_or_Hispanic': 0,
+        'Native_Amer':0,
+        'Multiple_Race':0,
+        'Other': 0,
+        'Pacific_Islander': 0,
+        'White': 0,
+        'Unknown': 0
```
Comment (Mr0grog) [inline covid19_sfbayarea/data/sonoma.py:207]: Seems like the base value for `African_Amer`, `Native_Amer`, `Multiple_Race`, and `Pacific_Islander` should probably be either `-1` or not present here to clarify that the counts are unknown rather than 0 (since they’re included in `Asian` or `Other`).

File: covid19_sfbayarea/data/sonoma.py [hunk 27]
```diff
@@ -0,0 +1,261 @@
...
+    Takes in a BeautifulSoup tag for a table and returns a list of
+    dictionaries in which the keys are strings and the values integers
+    """
+    categories: List[Dict[str, Union[str, int]]] = []
+    rows = get_rows(tag)
+    for row in rows:
+        group, cases, *rest = get_cells(row)
+        raw_cases = parse_int(cases)
+
+        element: Dict[str, Union[str, int]] = {'group': group, 'raw_cases': raw_cases}
```
Comment (Mr0grog) [inline covid19_sfbayarea/data/sonoma.py:175]: This should be `raw_count`:

```suggestion
        element: Dict[str, Union[str, int]] = {'group': group, 'raw_count': raw_cases}
```

File: covid19_sfbayarea/data/sonoma.py [hunk 28]
```diff
@@ -0,0 +1,261 @@
...
+    cumul_cases = 0
+    deaths = []
+    cumul_deaths = 0
+    # recovered = []
+    # cumul_recovered = 0
+    # active = []
+    # cumul_active = 0
+    rows = reversed(get_rows(cases_tag))
+    for row in rows:
+        row_cells = row.find_all(['th', 'td'])
```
[the remark this answers is not in the document]
Comment (ldtcooper) [reply covid19_sfbayarea/data/sonoma.py]: I think there was, but it seems to be working using `get_cells` so I changed it

File: README.md (hunk 13 above)
Comment (Mr0grog) [reply README.md:12]: So… we’re just going to put this in here as-is?

File: covid19_sfbayarea/data/sonoma.py [hunk 29]
```diff
@@ -0,0 +1,261 @@
...
+def row_list_to_dict(row: List[str], headers: List[str]) -> Dict[str, str]:
+    return dict(zip(headers, row))
+
+def parse_table(tag: element.Tag) -> List[Dict[str, str]]:
+    rows = tag.find_all('tr')
+    header = rows[0]
+    body = rows[1:]
+    header_cells = get_cells(header)
+    body_cells = (get_cells(row) for row in body)
+    return [row_list_to_dict(row, header_cells) for row in body_cells]
```
[the remark this answers is not in the document]
Comment (Mr0grog) [reply covid19_sfbayarea/data/sonoma.py:52]: Since you asked for a re-review, does this mean you’re going to keep these?

File: covid19_sfbayarea/data/sonoma.py [hunk 30]
```diff
@@ -0,0 +1,261 @@
...
+    """
+    Generates a timestamp string (e.g. May 6, 2020 10:00 AM) for when the scraper is run
+    """
+    update_time_text = soup.find('time', {'class': 'updated'}).text.strip()
+    try:
+        date = dateutil.parser.parse(update_time_text)
+    except ValueError:
+        raise ValueError(f'Date is not in ISO 8601'
+                         f'format: "{update_time_text}"')
+    return date.isoformat()
```
[the remark this answers is not in the document]
Comment (Mr0grog) [reply covid19_sfbayarea/data/sonoma.py:71]: Keeping this without timezone info?

**Comment by ldtcooper** [comment]
I did just miss those. Let me take a look

File: covid19_sfbayarea/data/sonoma.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (ldtcooper) [reply covid19_sfbayarea/data/sonoma.py]: I ended up redoing all the transform functions to use the new functions

File: covid19_sfbayarea/data/sonoma.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (ldtcooper) [reply covid19_sfbayarea/data/sonoma.py]: I went with the `datetime` solution to get this working

**Comment by ldtcooper** [comment]
Okay, I think I addressed those three points

File: covid19_sfbayarea/data/sonoma.py [hunk 31]
```diff
@@ -0,0 +1,261 @@
...
+    # this lets us get the second cases table
+    return tables[-1]
+
+def get_cells(row: element.ResultSet) -> List[str]:
+    """
+    Gets all th and tr elements within a single tr element
+    """
+    return [el.text for el in row.find_all(['th', 'td'])]
+
+def row_list_to_dict(row: List[str], headers: List[str]) -> UnformattedSeriesItem:
```
Comment (Mr0grog) [inline covid19_sfbayarea/data/sonoma.py:39]: Minor nit: one of the important things about naming (or aliasing) types like this is to change how you conceptualize your values and functions (e.g. you shouldn’t be thinking of `UnformattedSeriesItem` like a shortcut for `Dict[str, str]` here; you should be thinking of it like a subclass of `dict` — it should conceptually be its own separate thing).

So if you’re changing the return type to something named `UnformattedSeriesItem`, it’s probably a good idea to change the function name to not talk about making a `dict` and instead call it something like `row_list_to_series_item` or something.

File: covid19_sfbayarea/data/sonoma.py [hunk 32]
```diff
@@ -0,0 +1,261 @@
...
+    Where each list contains dictionaries (representing each day's data)
+    of form (example for cases):
+    { 'date': '', 'cases': -1, 'cumul_cases': -1 }
+    """
+    cases = []
+    cumul_cases = 0
+    deaths = []
+    cumul_deaths = 0
+
+    rows = list(reversed(parse_table(cases_tag)))
```
Comment (Mr0grog) [inline covid19_sfbayarea/data/sonoma.py:103]: Oh, forgot to comment on this in the review: I don’t think there’s any reason to convert this to a list, since you’re only iterating over it once and not returning it:

```suggestion
    rows = reversed(parse_table(cases_tag))
```
